pkg/kubelet/pluginmanager/cache/actual_state_of_world.go

```diff
@@ -0,0 +1,127 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+Package cache implements data structures used by the kubelet plugin manager to
+keep track of registered plugins.
+*/
+package cache
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/klog"
+)
+
+// ActualStateOfWorld defines a set of thread-safe operations for the kubelet
+// plugin manager's actual state of the world cache.
+// This cache contains a map of socket file path to plugin information of
+// all plugins attached to this node.
+type ActualStateOfWorld interface {
+
+	// GetRegisteredPlugins generates and returns a list of plugins
+	// that are successfully registered plugins in the current actual state of world.
+	GetRegisteredPlugins() []PluginInfo
+
+	// AddPlugin add the given plugin in the cache.
+	// An error will be returned if socketPath of the PluginInfo object is empty.
+	// Note that this is different from desired world cache's AddOrUpdatePlugin
+	// because for the actual state of world cache, there won't be a scenario where
+	// we need to update an existing plugin if the timestamps don't match. This is
+	// because the plugin should have been unregistered in the reconciller and therefore
+	// removed from the actual state of world cache first before adding it back into
+	// the actual state of world cache again with the new timestamp
+	AddPlugin(pluginInfo PluginInfo) error
+
+	// RemovePlugin deletes the plugin with the given socket path from the actual
+	// state of world.
+	// If a plugin does not exist with the given socket path, this is a no-op.
+	RemovePlugin(socketPath string)
+
+	// PluginExists checks if the given plugin exists in the current actual
+	// state of world cache with the correct timestamp
+	PluginExistsWithCorrectTimestamp(pluginInfo PluginInfo) bool
+}
+
+// NewActualStateOfWorld returns a new instance of ActualStateOfWorld
+func NewActualStateOfWorld() ActualStateOfWorld {
+	return &actualStateOfWorld{
+		socketFileToInfo: make(map[string]PluginInfo),
+	}
+}
+
+type actualStateOfWorld struct {
+
+	// socketFileToInfo is a map containing the set of successfully registered plugins
+	// The keys are plugin socket file paths. The values are PluginInfo objects
+	socketFileToInfo map[string]PluginInfo
+	sync.RWMutex
+}
+
+var _ ActualStateOfWorld = &actualStateOfWorld{}
+
+// PluginInfo holds information of a plugin
+type PluginInfo struct {
+	SocketPath           string
+	FoundInDeprecatedDir bool
+	Timestamp            time.Time
```

Comment:
This is super nit picky, but this code is so well factored, this might be worth fixing (it will discourage others from misusing it in the future).

`Timestamp` is an internal implementation detail of actual and desired state. But because we require external callers to pass in the `PluginInfo` object they are forced to set it. Instead, it would be much cleaner if we did the following:

1. Change `desiredStateOfWorld.AddOrUpdatePlugin(pluginInfo PluginInfo)` to only take the required parameters: `desiredStateOfWorld.AddOrUpdatePlugin(socketPath string, foundInDeprecatedDir bool)` (no `timestamp`).
2. Change the implementation of `desiredStateOfWorld.AddOrUpdatePlugin(socketPath string, foundInDeprecatedDir bool)` to automatically set `plugininfo.Timestamp` to `time.Now()` for `dsw.socketFileToInfo[socketPath]`.
3. Update the callers for `desiredStateOfWorld.AddOrUpdatePlugin(...)` so they no longer have to pass the full `PluginInfo` object (or the `Timestamp`).

Also, if you're feeling really inspired and want to get really fancy with separation of concerns, we could also rename `PluginInfo` to `pluginInfo` to make it [private](https://golang.org/ref/spec#Exported_identifiers) to the `cache` package. But that's a much bigger change, because it is being used all over the place.

All of this is totally optional, if you want to, if not, no worries.